solidity/contracts/bridge/Wallets.sol

```diff
@@ -456,6 +464,115 @@ library Wallets {
+    ///        not directly stored in the contract for gas efficiency purposes
+    ///        but they can be read from appropriate `DkgResultSubmitted`
+    ///        and `DkgResultApproved` events.
+    ///      - The `walletMemberIndex` must be in range [1, walletMembersIDs.length]
+    ///      - The caller must be the member of the source wallet signing group
+    ///        at the position indicated by `walletMemberIndex` parameter
+    ///      - The `walletMainUtxo` components must point to the recent main
+    ///        UTXO of the source wallet, as currently known on the Ethereum
+    ///        chain. If the source wallet has no main UTXO, this parameter
+    ///        can be empty as it is ignored since the wallet balance is
+    ///        assumed to be zero.
+    ///      - At least one Live wallet must exist in the system
+    ///      - Submitted target wallets count must match the expected count
+    ///        `N = min(liveWalletsCount, ceil(walletBtcBalance / walletMaxBtcTransfer))`
+    ///        where `N > 0`
+    ///      - Each target wallet must be not equal to the source wallet
+    ///      - Each target wallet must be in Live state
+    function notifyWalletMovingFundsCommitmentSubmitted(
+        BridgeState.Storage storage self,
+        bytes20 walletPubKeyHash,
+        BitcoinTx.UTXO calldata walletMainUtxo,
+        uint32[] calldata walletMembersIDs,
+        uint256 walletMemberIndex,
+        bytes20[] calldata targetWallets
+    ) internal {
+        Wallet storage wallet = self.registeredWallets[walletPubKeyHash];
+
+        require(
+            wallet.state == WalletState.MovingFunds,
+            "Source wallet must be in MovingFunds state"
+        );
+
+        require(
+            wallet.pendingRedemptionsValue == 0,
+            "Source wallet must handle all pending redemptions first"
+        );
+
+        require(
+            wallet.movingFundsTargetWalletsCommitmentHash == bytes32(0),
+            "Target wallets commitment already submitted"
+        );
+
+        require(
+            self.ecdsaWalletRegistry.isWalletMember(
+                wallet.ecdsaWalletID,
+                walletMembersIDs,
+                msg.sender,
+                walletMemberIndex
+            ),
+            "Caller is not a member of the source wallet"
+        );
+
+        uint64 walletBtcBalance = getWalletBtcBalance(
+            self,
+            walletPubKeyHash,
+            walletMainUtxo
+        );
+        uint256 expectedTargetWalletsCount = Math.min(
+            self.liveWalletsCount,
+            Math.ceilDiv(walletBtcBalance, self.walletMaxBtcTransfer)
+        );
+
+        // This requirement fails only when `liveWalletsCount` is zero. In
+        // that case, the system cannot accept the commitment and must provide
+        // new wallets first.
+        require(expectedTargetWalletsCount > 0, "No target wallets available");
```

Comment:
The active wallet can not be notified as closeable (see `notifyCloseableWallet`) but it can fail the heartbeat and transition into `MovingFunds` state. If there is no other wallet in the system, the active wallet can not move its funds.

We end up with a scenario when we have one `MovingFunds` wallet in the system, not accepting new deposits, not accepting redemption requests, so there is no way to increase or decrease its balance, and there is no way to create a new wallet in another way than by governance action lowering `walletMaxBtcBalance`. There is a way to fix it by the governance but it is not ideal. What is worse though is that it complicates timeouts for completing the moving funds operation. If there are no live wallets in the system, and the wallet can not commit to target wallets, the time for moving funds should not start counting.

A similar scenario is when the active wallet failed the heartbeat and transitioned into `MovingFunds` state and some other wallet qualified as closeable, entered the `MovingFunds` state, and has no `Live` wallets in the system it could move the funds to. 

One idea I had would be to disallow the active wallet to fail the heartbeat but it does not solve the problem given that nothing stops the active wallet to enter `MovingFunds` state by a timed-out redemption.

The only solution I see is to alter the algorithm for requesting a new wallet, and let the new wallet creation to be requested if the active wallet is in `MovingFunds` state, no matter its age or balance. What do you think? cc @beaushinkle @michalinacienciala

Reply:
> (...) there is no way to create a new wallet in another way than by governance action lowering walletMaxBtcBalance.

This is not true. If an active wallet goes `MovingFunds`, the active wallet slot gets cleared out and a new wallet creation can be requested immediately (see `Wallets.requestNewWallet`). That basically solves the problem you described.

Reply:
> We end up with a scenario when we have one MovingFunds wallet in the system, not accepting new deposits, not accepting redemption requests, so there is no way to increase or decrease its balance, and there is no way to create a new wallet in another way than by governance action lowering walletMaxBtcBalance

> The only solution I see is to alter the algorithm for requesting a new wallet, and let the new wallet creation to be requested if the active wallet is in MovingFunds state, no matter its age or balance

My understanding is that when the wallet marked as 'active' goes to `MovingFunds ` state, the 'active' mark gets removed (`delete self.activeWalletPubKeyHash` in `moveFunds` function). And the function for creating a new wallet allows for creation of the wallet when there's no 'active' wallet set.
Am I missing something?

Reply:
Oh, I see Łukasz wrote basically the same :)

Reply:
> The active wallet can not be notified as closeable (see notifyCloseableWallet) but it can fail the heartbeat and transition into MovingFunds state. If there is no other wallet in the system, the active wallet can not move its funds.

> We end up with a scenario when we have one MovingFunds wallet in the system, not accepting new deposits, not accepting redemption requests, so there is no way to increase or decrease its balance, and there is no way to create a new wallet in another way than by governance action lowering walletMaxBtcBalance. There is a way to fix it by the governance but it is not ideal. What is worse though is that it complicates timeouts for completing the moving funds operation. If there are no live wallets in the system, and the wallet can not commit to target wallets, the time for moving funds should not start counting.

I think the simple solution is that a wallet should only be allowed to go into the `MovingFunds` state if there's a wallet for it to move its funds to, regardless of circumstance. To accomplish this, we could make all of the `intention-to-close` transactions (including the one we send for failed heartbeats) include a live wallet that we validate as being live at the time.

Reply:
`requestNewWallet` seems like it would take too long (iirc, it take 4-5 days), and by then the timeout for wallet transfer would have probably elapsed unless we make that timeout *very* long.

If we make it so that a wallet can't enter the `MovingFunds` state without another live wallet in the system, as in all `intention-to-close` transactions require a live wallet, then that solves heartbeats and whatnot, but doesn't solve timed out redemptions.

For timed out redemptions, what's the ideal behavior? There's a live wallet on the system, it'll be days before we can get a new one, it just failed to redeem, and it has some amount of bitcoin. What do we want it to do?

1) continue to accept deposits or stop accepting deposits
2) continue trying to serve redemptions or stop serving redemptions

If we want it to to continue, we slash for the failed redemption but don't go through with the closing process. If we want it to stop, we put it in the `MovingFunds` state with it's own, longer timeout. Maybe this looks like a new state called `MovingFundsToNewWallet` or something with a 7.

Reply:
We discussed this problem on chat and agreed to expose a separate function allowing to reset the timeout for moving funds in case there is no active wallet in the system.

Example:
- Wallet transitioned to MovingFunds state.
- The wallet wants to commit to target wallets but there is no single Live wallet in the system. There are 4 days before the timeout for moving funds is hit.
- After 2 days, there are still no Live wallets. The wallet is calling a function to reset moving funds timeout. The wallet has again 4 days to move funds.

@lukasz-zimnoch Can we leave a TODO in `MovingFunds` library?

Reply:
Done.